### PR TITLE
integrate the mDNS client into the publisher. Start mDNS client, if there is a existing server running on the system.

### PR DESCRIPTION
**Pull request by LeZhang2016**

### CODE REVIEW

File: src/web/mdns-publisher/mdns_publisher.cpp [hunk 1]
```diff
@@ -95,86 +190,147 @@ int Publisher::CreateService(AvahiServer *aServer)
...
+        if (!mServerGroup)
             CreateService(aServer);
         break;
 
     case AVAHI_SERVER_COLLISION:
-        char *newHostName;
+        avahi_simple_poll_quit(mSimplePoll);
+        if (mServer)
+            avahi_server_free(mServer);
+        StartClient();
```
Comment (bukepo) [inline src/web/mdns-publisher/mdns_publisher.cpp]: missing `break;` below?

File: src/web/mdns-publisher/mdns_publisher.hpp [hunk 2]
```diff
@@ -54,33 +56,60 @@ namespace Mdns {
...
+
+    static void ClientCallback(AvahiClient *aClient, AvahiClientState aState,
+                               AVAHI_GCC_UNUSED void *aUserData);
+    void ClientCallback(AvahiClient *aClient, AvahiClientState aState);
+
+    static void PeriodicallyPublish(AVAHI_GCC_UNUSED AvahiTimeout *aTimeout, AVAHI_GCC_UNUSED void *aUserData);
+    void PeriodicallyPublish(AVAHI_GCC_UNUSED AvahiTimeout *aTimeout);
+    void Free(void);
+
+    static Publisher &Instance(void)
```
Comment (jwhui) [inline src/web/mdns-publisher/mdns_publisher.hpp]: Methods should have a verb (e.g. `GetInstance()`).

File: src/web/mdns-publisher/mdns_publisher.cpp [hunk 3]
```diff
@@ -31,6 +31,9 @@
  *   This file implements starting an mDNS server, and publish border router service.
  */
 
+#include <avahi-client/publish.h>
+#include <avahi-core/publish.h>
+
 #include "common/code_utils.hpp"
 
 #include "mdns_publisher.hpp"
```
Comment (jwhui) [inline src/web/mdns-publisher/mdns_publisher.cpp]: The compilation unit's corresponding header should appear first.

File: src/web/mdns-publisher/mdns_publisher.cpp [hunk 4]
```diff
@@ -95,86 +191,147 @@ int Publisher::CreateService(AvahiServer *aServer)
...
+    case AVAHI_CLIENT_S_COLLISION:
+    case AVAHI_CLIENT_S_REGISTERING:
+
+        if (mClientGroup)
+            avahi_entry_group_reset(mClientGroup);
+
+        break;
+
+    case AVAHI_CLIENT_CONNECTING:
+        ;
```
Comment (jwhui) [inline src/web/mdns-publisher/mdns_publisher.cpp]: I know `break` is not needed, but it is cleaner to include it.

File: src/web/mdns-publisher/mdns_publisher.cpp [hunk 5]
```diff
@@ -95,86 +191,147 @@ int Publisher::CreateService(AvahiServer *aServer)
...
     case AVAHI_SERVER_FAILURE:
 
         syslog(LOG_ERR, "Server failure: %s\n",
                avahi_strerror(avahi_server_errno(aServer)));
         avahi_simple_poll_quit(mSimplePoll);
-        isStart = false;
         break;
 
     case AVAHI_SERVER_INVALID:
         ;
```
Comment (jwhui) [inline src/web/mdns-publisher/mdns_publisher.cpp]: Include `break` here.

File: src/web/mdns-publisher/mdns_publisher.cpp [hunk 6]
```diff
@@ -95,86 +191,147 @@ int Publisher::CreateService(AvahiServer *aServer)
...
             avahi_simple_poll_get(mSimplePoll),
             avahi_elapse_time(&tv, PERIODICAL_TIME, 0),
             PeriodicallyPublish,
-            mServer);
+            this);
+    }
+
+    if ((mClient != NULL) && (avahi_client_get_state(mClient) == AVAHI_CLIENT_S_RUNNING))
+    {
+
```
Comment (jwhui) [inline src/web/mdns-publisher/mdns_publisher.cpp]: Unnecessary newline.

File: src/web/mdns-publisher/mdns_publisher.cpp [hunk 7]
```diff
@@ -95,86 +191,147 @@ int Publisher::CreateService(AvahiServer *aServer)
...
+        avahi_elapse_time(&tv, PERIODICAL_TIME, 0),
+        PeriodicallyPublish,
+        this);
+    avahi_simple_poll_loop(mSimplePoll);
+
+exit:
+
+    Free();
+    return ret;
+
```
Comment (jwhui) [inline src/web/mdns-publisher/mdns_publisher.cpp]: Unnecessary newline.

File: src/web/mdns-publisher/mdns_publisher.cpp [hunk 8]
```diff
@@ -207,35 +363,73 @@ int Publisher::StartServer(void)
...
+        syslog(LOG_ERR, "Entry group failure: %s\n",
+               avahi_strerror(avahi_client_errno(avahi_entry_group_get_client(aGroup))));
+
+        avahi_simple_poll_quit(mSimplePoll);
+        break;
+
+    case AVAHI_ENTRY_GROUP_UNCOMMITED:
+    case AVAHI_ENTRY_GROUP_REGISTERING:
+        ;
```
Comment (jwhui) [inline src/web/mdns-publisher/mdns_publisher.cpp]: Include `break`.

File: src/web/mdns-publisher/mdns_publisher.hpp [hunk 9]
```diff
@@ -54,33 +56,60 @@ namespace Mdns {
...
-    static AvahiSEntryGroup *mGroup;
-    static AvahiSimplePoll  *mSimplePoll;
-    static AvahiServer      *mServer;
-    static const char       *mNetworkNameTxt;
-    static const char       *mExtPanIdTxt;
-    static const char       *mType;
-    static uint16_t          mPort;
-    static char             *mServiceName;
-    static bool              isStart;
+    void ServerCallback(AvahiServer *aServer, AvahiServerState aState);
```
Comment (jwhui) [inline src/web/mdns-publisher/mdns_publisher.hpp]: Methods should include verbs.  For example `HandleServerCallback()`, or something more appropriate.

File: src/web/mdns-publisher/mdns_publisher.hpp [hunk 10]
```diff
@@ -54,33 +56,60 @@ namespace Mdns {
...
+
+    static void ClientCallback(AvahiClient *aClient, AvahiClientState aState,
+                               AVAHI_GCC_UNUSED void *aUserData);
+    void ClientCallback(AvahiClient *aClient, AvahiClientState aState);
+
+    static void PeriodicallyPublish(AVAHI_GCC_UNUSED AvahiTimeout *aTimeout, AVAHI_GCC_UNUSED void *aUserData);
```
Comment (jwhui) [inline src/web/mdns-publisher/mdns_publisher.hpp]: Methods should include verbs.  `PeriodicallyPublish` does not tell me what this method actually does.

File: src/web/mdns-publisher/mdns_publisher.cpp [hunk 11]
```diff
@@ -78,21 +114,80 @@ void Publisher::Free(void)
...
+
+    if (!mClientGroup)
+    {
+        mClientGroup = avahi_entry_group_new(aClient, HandleEntryGroupStart, this);
+        VerifyOrExit(mClientGroup != NULL, ret = kMdnsPublisher_FailedCreateGoup);
+    }
+
+    if (avahi_entry_group_is_empty(mClientGroup))
+    {
+
```
Comment (LeZhang2016) [inline src/web/mdns-publisher/mdns_publisher.cpp]: space line

File: src/web/mdns-publisher/mdns_publisher.cpp [hunk 12]
```diff
@@ -260,35 +448,37 @@ void Publisher::EntryGroupCallback(AvahiServer *aServer,
...
-        ;
+        break;
     }
 }
 
-void Publisher::SetNetworkNameTxt(const char *aTxtData)
+void Publisher::SetNetworkNameTxt(const char *aNetworkNameTxt)
 {
-    mNetworkNameTxt = aTxtData;
+
```
Comment (bukepo) [inline src/web/mdns-publisher/mdns_publisher.cpp]: unnecessary new line.

File: src/web/mdns-publisher/mdns_publisher.cpp [hunk 13]
```diff
@@ -260,35 +448,37 @@ void Publisher::EntryGroupCallback(AvahiServer *aServer,
...
+    avahi_free(mNetworkNameTxt);
```
Comment (bukepo) [inline src/web/mdns-publisher/mdns_publisher.cpp]: `mNetworkNameTxt` is not initialized?

File: src/web/mdns-publisher/mdns_publisher.cpp [hunk 14]
```diff
@@ -260,35 +448,37 @@ void Publisher::EntryGroupCallback(AvahiServer *aServer,
...
+    mNetworkNameTxt = avahi_strdup(aNetworkNameTxt);
 }
 
-void Publisher::SetExtPanIdTxt(const char *aTxtData)
+void Publisher::SetExtPanIdTxt(const char *aExtPanIdTxt)
 {
-    mExtPanIdTxt = aTxtData;
+    avahi_free(mExtPanIdTxt);
```
Comment (bukepo) [inline src/web/mdns-publisher/mdns_publisher.cpp]: not initialized?

File: src/web/mdns-publisher/mdns_publisher.cpp [hunk 15]
```diff
@@ -260,35 +448,37 @@ void Publisher::EntryGroupCallback(AvahiServer *aServer,
...
+    mExtPanIdTxt = avahi_strdup(aExtPanIdTxt);
 }
 
 void Publisher::SetType(const char *aType)
 {
-
-    mType = aType;
+    avahi_free(mType);
```
Comment (bukepo) [inline src/web/mdns-publisher/mdns_publisher.cpp]: not initialized?

File: script/bootstrap [hunk 16]
```diff
@@ -44,7 +44,7 @@ install_packages_apt()
     sudo apt-get install -y autoconf-archive
 
     # mDNS
-    sudo apt-get install -y libavahi-core-dev
+    sudo apt-get install -y libavahi-common-dev libavahi-client-dev
```
Comment (bukepo) [inline script/bootstrap]: Is `avahi-daemon` required?

File: src/web/mdns-publisher/mdns_publisher.cpp [hunk 17]
```diff
@@ -28,267 +28,277 @@
...
 
-    if (mSimplePoll)
+    if (mSimplePoll != NULL)
     {
         avahi_simple_poll_free(mSimplePoll);
+        mSimplePoll = NULL;
     }
 
-    avahi_free(mServiceName);
+    if (mServiceName)
```
Comment (bukepo) [inline src/web/mdns-publisher/mdns_publisher.cpp]: Keep same style of checking pointer is NULL or not? I found
```cpp
if (xxx)
if (!xxx)
if (xxx != NULL)
(xxx == NULL)
```

File: src/web/mdns-publisher/mdns_publisher.cpp [hunk 18]
```diff
@@ -28,267 +28,277 @@
...
+    {
+        avahi_free(mServiceName);
+        mServiceName = NULL;
+    }
+
+    if (mClientGroup != NULL)
+    {
+        avahi_entry_group_free(mClientGroup);
```
Comment (bukepo) [inline src/web/mdns-publisher/mdns_publisher.cpp]: set it to NULL?

File: src/web/mdns-publisher/mdns_publisher.cpp [hunk 19]
```diff
@@ -28,267 +28,277 @@
...
+    {
+        syslog(LOG_ERR, "Adding service '%s'", mServiceName);
 
-    VerifyOrExit(avahi_server_add_service(aServer, mGroup, AVAHI_IF_UNSPEC,
-                                          AVAHI_PROTO_UNSPEC, (AvahiPublishFlags) 0,
-                                          mServiceName, mType, NULL, NULL, mPort,
-                                          mNetworkNameTxt, mExtPanIdTxt, NULL) == 0,
-                 ret = kMdnsPublisher_FailedAddSevice);
+        VerifyOrExit(avahi_entry_group_add_service(mClientGroup, AVAHI_IF_UNSPEC,
+                                                   AVAHI_PROTO_UNSPEC, (AvahiPublishFlags) 0,
```
Comment (bukepo) [inline src/web/mdns-publisher/mdns_publisher.cpp]: Use `static_cast<AvahiPublishFlags>`?

File: src/web/mdns-publisher/mdns_publisher.cpp [hunk 20]
```diff
@@ -28,267 +28,277 @@
...
+                                                   mServiceName, mType, NULL, NULL, mPort,
+                                                   mNetworkNameTxt, mExtPanIdTxt, NULL) == 0,
+                     ret = kMdnsPublisher_FailedAddSevice);
 
-    syslog(LOG_INFO, " Service Name: %s Port: %d Network Name: %s Extended Pan ID: %s ",
-           mServiceName, mPort, mNetworkNameTxt, mExtPanIdTxt);
+        syslog(LOG_INFO, " Service Name: %s  Port: %d  Network Name: %s  Extended Pan ID: %s ",
```
Comment (bukepo) [inline src/web/mdns-publisher/mdns_publisher.cpp]: unnecessary space at end of syslog format string. And is the preceding space in intended?

File: src/web/mdns-publisher/mdns_publisher.cpp [hunk 21]
```diff
@@ -28,267 +28,277 @@
...
-                 ret = kMdnsPublisher_FailedRegisterSevice);
-    return ret;
+        VerifyOrExit(avahi_entry_group_commit(mClientGroup) == 0,
+                     ret = kMdnsPublisher_FailedRegisterSevice);
+    }
 
 exit:
-    if (ret != kMdnsPublisher_OK)
+
+    if (ret == kMdnsPublisher_FailedAddSevice)
```
Comment (bukepo) [inline src/web/mdns-publisher/mdns_publisher.cpp]: use switch?

File: src/web/mdns-publisher/mdns_publisher.cpp [hunk 22]
```diff
@@ -28,267 +28,277 @@
...
+    {
+        serviceName = avahi_alternative_service_name(mServiceName);
+        avahi_free(mServiceName);
+        mServiceName = serviceName;
+        syslog(LOG_INFO, "Service name collision, renaming service to '%s'", mServiceName);
+        avahi_entry_group_reset(mClientGroup);
+        CreateService(aClient);
```
Comment (bukepo) [inline src/web/mdns-publisher/mdns_publisher.cpp]: Is it intended to ignore error from `CreateService()`?

File: src/web/mdns-publisher/mdns_publisher.cpp [hunk 23]
```diff
@@ -28,267 +28,277 @@
...
 
-    case AVAHI_SERVER_INVALID:
-        ;
-    }
-exit:
-    if (ret < 0)
-    {
-        avahi_simple_poll_quit(mSimplePoll);
+    case AVAHI_CLIENT_CONNECTING:
+        break;
```
Comment (bukepo) [inline src/web/mdns-publisher/mdns_publisher.cpp]: missing `default` case?

**Comment by LeZhang2016** [comment]
@bukepo, Thanks a lot for your comments and suggestions, I have updated this RP according to your comments.

**Pull request merged**